Add continuous time wrappers

bnham opened the pull request:
#### ce9238afa421a5eefce7c51d5c627a8b7eacaf79
<pre>
Add continuous time wrappers
<a href="https://bugs.webkit.org/show_bug.cgi?id=285564">https://bugs.webkit.org/show_bug.cgi?id=285564</a>
<a href="https://rdar.apple.com/142512176">rdar://142512176</a>

Reviewed by Basuke Suzuki.

It&apos;s useful to have a time type which advances while the system is asleep. `MonotonicTime` does not
guarantee this (and definitely doesn&apos;t advance while the system is asleep on Darwin). To that end,
this adds two such time types: `ContinuousTime` and `ContinuousApproximateTime`.

- On Darwin, these are wrappers for `mach_continuous_time` and `mach_continuous_approximate_time`.
- On Linux and OpenBSD, both are wrappers for `CLOCK_BOOTTIME`, which advances while the system is
  asleep (there doesn&apos;t seem to be an &quot;approximate&quot; faster version of that clock).
- On other systems, these just wrap the wall clock, with some protection against time going
  backwards. This is what we were already doing for `MonotonicTime` on some platforms.

We also change the two call sites where we were using continuous time:

- `SystemTracing` was using `mach_continuous_time` since it&apos;s built on top of `os_signpost`, which
  itself expects events in `mach_continuous_time` units. Move this to using `ContinuousTime`.
- `WebPrivacyHelpers` was using `CLOCK_MONOTONIC_RAW_APPROX` (which is just
  `mach_continuous_approximate_time` under the hood). Move this to using
  `ContinuousApproximateTime`.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/ClockType.cpp:
(WTF::printInternal):
* Source/WTF/wtf/ClockType.h:
* Source/WTF/wtf/ContinuousApproximateTime.cpp: Copied from Source/WTF/wtf/ClockType.cpp.
(WTF::ContinuousApproximateTime::approximateWallTime const):
(WTF::ContinuousApproximateTime::approximateMonotonicTime const):
(WTF::ContinuousApproximateTime::dump const):
* Source/WTF/wtf/ContinuousApproximateTime.h: Added.
(WTF::ContinuousApproximateTime::MarkableTraits::isEmptyValue):
(WTF::ContinuousApproximateTime::MarkableTraits::emptyValue):
* Source/WTF/wtf/ContinuousTime.cpp: Copied from Source/WTF/wtf/ClockType.cpp.
(WTF::ContinuousTime::approximateWallTime const):
(WTF::ContinuousTime::approximateMonotonicTime const):
(WTF::ContinuousTime::dump const):
* Source/WTF/wtf/ContinuousTime.h: Added.
(WTF::ContinuousTime::MarkableTraits::isEmptyValue):
(WTF::ContinuousTime::MarkableTraits::emptyValue):
* Source/WTF/wtf/CurrentTime.cpp:
* Source/WTF/wtf/Seconds.cpp:
(WTF::Seconds::operator+ const):
(WTF::Seconds::operator- const):
* Source/WTF/wtf/Seconds.h:
* Source/WTF/wtf/TimeWithDynamicClockType.cpp:
(WTF::TimeWithDynamicClockType::now):
(WTF::TimeWithDynamicClockType::continuousTime const):
(WTF::TimeWithDynamicClockType::continuousApproximateTime const):
(WTF::TimeWithDynamicClockType::approximateWallTime const):
(WTF::TimeWithDynamicClockType::approximateMonotonicTime const):
* Source/WTF/wtf/TimeWithDynamicClockType.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFCurrentContinuousTime):
(machTimebaseInfo): Deleted.
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::RestrictedOpenerDomainsController::scheduleNextUpdate):
(WebKit::RestrictedOpenerDomainsController::lookup const):
(WebKit::approximateContinuousTimeNanoseconds): Deleted.

Canonical link: <a href="https://commits.webkit.org/288621@main">https://commits.webkit.org/288621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10e7a0232b6e5ec44d1ecd6a0b05de3776468c95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83907 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38208 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88981 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34915 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3616 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11492 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65265 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23100 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45557 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2611 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30461 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33964 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/76870 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73600 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31215 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90357 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/82924 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11172 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8107 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73719 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11396 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72068 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72937 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18041 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17214 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2506 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11124 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16596 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105342 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10972 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25458 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14448 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12744 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

basuke commented on Source/WTF/wtf/ContinuousApproximateTime.h:
```diff
@@ -0,0 +1,83 @@
+// ContinuousTime::now().
+class ContinuousApproximateTime final : public GenericTimeMixin<ContinuousApproximateTime> {
+public:
+    static constexpr ClockType clockType = ClockType::Continuous;
```
Isn't this `ClockType::ContinuousApproximate`?

bnham replied:
Oops that's correct, will fix

basuke commented on Source/WTF/wtf/CurrentTime.cpp:
```diff
@@ -314,4 +340,40 @@ ApproximateTime ApproximateTime::now()
+    if (currentTimeNow < lastTime)
+        return lastTime;
+    lastTime = currentTimeNow;
+    return fromRawSeconds(currentTimeNow);
```
These logic is now reused for three times. We should extract this like `monotonicCurrentTime()` or something and use that to pass the result to `fromRawSeconds()`.

basuke commented on Source/WTF/wtf/CurrentTime.cpp:
```diff
@@ -314,4 +340,40 @@ ApproximateTime ApproximateTime::now()
+    if (currentTimeNow < lastTime)
+        return lastTime;
+    lastTime = currentTimeNow;
+    return fromRawSeconds(currentTimeNow);
```
Ditto.

basuke reviewed (approved):
r+ with a comment for generic logic.

bnham committed:
Add continuous time wrappers
https://bugs.webkit.org/show_bug.cgi?id=285564
rdar://142512176

Reviewed by Basuke Suzuki.

It's useful to have a time type which advances while the system is asleep. `MonotonicTime` does not
guarantee this (and definitely doesn't advance while the system is asleep on Darwin). To that end,
this adds two such time types: `ContinuousTime` and `ContinuousApproximateTime`.

- On Darwin, these are wrappers for `mach_continuous_time` and `mach_continuous_approximate_time`.
- On Linux and OpenBSD, both are wrappers for `CLOCK_BOOTTIME`, which advances while the system is
  asleep (there doesn't seem to be an "approximate" faster version of that clock).
- On other systems, these just wrap the wall clock, with some protection against time going
  backwards. This is what we were already doing for `MonotonicTime` on some platforms.

We also change the two call sites where we were using continuous time:

- `SystemTracing` was using `mach_continuous_time` since it's built on top of `os_signpost`, which
  itself expects events in `mach_continuous_time` units. Move this to using `ContinuousTime`.
- `WebPrivacyHelpers` was using `CLOCK_MONOTONIC_RAW_APPROX` (which is just
  `mach_continuous_approximate_time` under the hood). Move this to using
  `ContinuousApproximateTime`.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/ClockType.cpp:
(WTF::printInternal):
* Source/WTF/wtf/ClockType.h:
* Source/WTF/wtf/ContinuousApproximateTime.cpp: Copied from Source/WTF/wtf/ClockType.cpp.
(WTF::ContinuousApproximateTime::approximateWallTime const):
(WTF::ContinuousApproximateTime::approximateMonotonicTime const):
(WTF::ContinuousApproximateTime::dump const):
* Source/WTF/wtf/ContinuousApproximateTime.h: Added.
(WTF::ContinuousApproximateTime::MarkableTraits::isEmptyValue):
(WTF::ContinuousApproximateTime::MarkableTraits::emptyValue):
* Source/WTF/wtf/ContinuousTime.cpp: Copied from Source/WTF/wtf/ClockType.cpp.
(WTF::ContinuousTime::approximateWallTime const):
(WTF::ContinuousTime::approximateMonotonicTime const):
(WTF::ContinuousTime::dump const):
* Source/WTF/wtf/ContinuousTime.h: Added.
(WTF::ContinuousTime::MarkableTraits::isEmptyValue):
(WTF::ContinuousTime::MarkableTraits::emptyValue):
* Source/WTF/wtf/CurrentTime.cpp:
* Source/WTF/wtf/Seconds.cpp:
(WTF::Seconds::operator+ const):
(WTF::Seconds::operator- const):
* Source/WTF/wtf/Seconds.h:
* Source/WTF/wtf/TimeWithDynamicClockType.cpp:
(WTF::TimeWithDynamicClockType::now):
(WTF::TimeWithDynamicClockType::continuousTime const):
(WTF::TimeWithDynamicClockType::continuousApproximateTime const):
(WTF::TimeWithDynamicClockType::approximateWallTime const):
(WTF::TimeWithDynamicClockType::approximateMonotonicTime const):
* Source/WTF/wtf/TimeWithDynamicClockType.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(WTFCurrentContinuousTime):
(machTimebaseInfo): Deleted.
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::RestrictedOpenerDomainsController::RestrictedOpenerDomainsController):
(WebKit::RestrictedOpenerDomainsController::scheduleNextUpdate):
(WebKit::RestrictedOpenerDomainsController::lookup const):
(WebKit::approximateContinuousTimeNanoseconds): Deleted.

Canonical link: https://commits.webkit.org/288621@main
